### PR TITLE
Fixes #2360 - Add support for manually marking rooms as unread

### LINKED ISSUE #2360: [Task] Manually mark a room as unread
Please see and discuss the details in the meta issue.

### PR DESCRIPTION
**Pull request by stefanceriu**
Requires https://github.com/matrix-org/matrix-rust-sdk/pull/3076

### CODE REVIEW

File: ElementX/Sources/Screens/HomeScreen/HomeScreenViewModel.swift [hunk 1]
```diff
@@ -143,6 +147,26 @@ class HomeScreenViewModel: HomeScreenViewModelType, HomeScreenViewModelProtocol
             actionsSubject.send(.presentInvitesScreen)
         case .globalSearch:
             actionsSubject.send(.presentGlobalSearch)
+        case .markRoomAsUnread(let roomIdentifier):
+            Task {
+                if let roomProxy = await userSession.clientProxy.roomForIdentifier(roomIdentifier) {
+                    if case .failure(let error) = await roomProxy.markAsUnread() {
```
Comment (Velin92) [inline ElementX/Sources/Screens/HomeScreen/HomeScreenViewModel.swift:153]: can we use `switch` here? if there must be an else... I would rather prefer it than `if case`.

File: ElementX/Sources/Screens/HomeScreen/HomeScreenViewModel.swift [hunk 2]
```diff
@@ -143,6 +147,26 @@ class HomeScreenViewModel: HomeScreenViewModelType, HomeScreenViewModelProtocol
...
+                        MXLog.error("Failed marking room \(roomIdentifier) as unread with error: \(error)")
+                    } else {
+                        ServiceLocator.shared.analytics.trackInteraction(name: .MobileRoomListRoomContextMenuUnreadToggle)
+                    }
+                }
+            }
+        case .markRoomAsRead(let roomIdentifier):
+            Task {
+                if let roomProxy = await userSession.clientProxy.roomForIdentifier(roomIdentifier) {
+                    if case .failure(let error) = await roomProxy.markAsRead(sendReadReceipts: true, receiptType: appSettings.sendReadReceiptsEnabled ? .read : .readPrivate) {
```
Comment (Velin92) [inline ElementX/Sources/Screens/HomeScreen/HomeScreenViewModel.swift:163]: same as above

File: ElementX.xcodeproj/project.xcworkspace/xcshareddata/swiftpm/Package.resolved [hunk 3]
```diff
@@ -121,17 +121,16 @@
       "kind" : "remoteSourceControl",
       "location" : "https://github.com/matrix-org/matrix-analytics-events",
       "state" : {
-        "revision" : "aa14cbcdf81af2746d20a71779ec751f971e1d7f",
-        "version" : "0.9.0"
+        "revision" : "28d1e65d306420b3affc32f719cb516f74a7406f"
```
Comment (Velin92) [inline ElementX.xcodeproj/project.xcworkspace/xcshareddata/swiftpm/Package.resolved:124]: I would have separated this analytics changes into another PR, I guess is fine now, but next time better have two smaller separated PRs

File: ElementX/Sources/Screens/HomeScreen/HomeScreenViewModel.swift (hunk 1 above)
Comment (stefanceriu) [reply ElementX/Sources/Screens/HomeScreen/HomeScreenViewModel.swift:153]: Good point 👍

File: ElementX.xcodeproj/project.xcworkspace/xcshareddata/swiftpm/Package.resolved (hunk 3 above)
Comment (stefanceriu) [reply ElementX.xcodeproj/project.xcworkspace/xcshareddata/swiftpm/Package.resolved:124]: That's fair, I just didn't feel like dealing with the conflicts 😅

File: ElementX/Sources/Screens/HomeScreen/View/HomeScreenRoomCell.swift [hunk 4]
```diff
@@ -133,29 +133,23 @@ struct HomeScreenRoomCell: View {
...
             }
         }
     }
-    
-    private var hasNewContent: Bool {
-        room.hasUnreadMessages ||
-            room.hasUnreadMentions ||
-            room.hasUnreadNotifications
-    }
-
```
Comment (Velin92) [inline ElementX/Sources/Screens/HomeScreen/View/HomeScreenRoomCell.swift:150]: thanks for the cleanup

File: changelog.d/2360.feature [hunk 5]
```diff
@@ -0,0 +1 @@
+Add support for manually marking a room as read/unread
```
Comment (Velin92) [inline changelog.d/2360.feature:1]: maybe let's also add a changelog about the analytics update?

**Review by Velin92 — APPROVED** [review]
LGTM, just a few suggestions

File: ElementX/Sources/Screens/HomeScreen/HomeScreenModels.swift [hunk 6]
```diff
@@ -171,10 +177,15 @@ struct HomeScreenRoom: Identifiable, Equatable {
...
     
+    var hasNewContent: Bool {
+        hasUnreadMessages || hasUnreadMentions || hasUnreadNotifications || isMarkedUnread
+    }
+    
     static func placeholder() -> HomeScreenRoom {
         HomeScreenRoom(id: UUID().uuidString,
                        roomId: nil,
                        name: "Placeholder room name",
+                       isMarkedUnread: false,
```
[the remark this answers is not in the document]
Comment (stefanceriu) [reply ElementX/Sources/Screens/HomeScreen/HomeScreenModels.swift:188]: It's basically the same as when you have a new message so I'm not sure that would bring any value, we already test all those cases

File: ElementX/Sources/Screens/HomeScreen/View/HomeScreenRoomCell.swift [hunk 7]
```diff
@@ -133,29 +133,23 @@ struct HomeScreenRoomCell: View {
...
     private var isHighlighted: Bool {
-        guard !room.isPlaceholder &&
-            room.notificationMode != .mute else {
+        guard !room.isPlaceholder && room.notificationMode != .mute else {
             return false
         }
-        return room.hasUnreadNotifications || room.hasUnreadMentions
-    }
         
+        return room.hasUnreadNotifications || room.hasUnreadMentions || room.isMarkedUnread
```
Comment (bmarty) [inline ElementX/Sources/Screens/HomeScreen/View/HomeScreenRoomCell.swift:150]: I understand that the `guard` will prevent a muted room from being highlighted when it's been manually marked as unread.

I think this is not correct from what I read [here](https://github.com/element-hq/element-meta/issues/2201):

> rooms marked as unread should show a green badge even if their notification settings are set to mentions&keywords or muted. After being marked as unread or being rendered a room should go back to showing the default colors for those options

Comment (stefanceriu) [reply ElementX/Sources/Screens/HomeScreen/View/HomeScreenRoomCell.swift:150]: You are absolutely right, good catch! I'll fix it asap.

**Pull request merged**